### ISSUE TITLE
After pausing in the reactor, animation loop doesn't resume

### ISSUE THREAD

**Issue by jwmerrill** [jwmerrill/elm-animation-frame#1]
https://groups.google.com/d/msg/elm-discuss/SLWhUe_KTwY/xr4YHe_BgCcJ

wighawag:

> Took the mario.elm from elm-examples and replaced (fps 30) with AnimationFrame.frame and get the following issue:
> when pausing with the elm-reactor and resuming (using the pause/play button), the mario character do not get updated anymore (it works fine with (fps 30)). 

Laszlo Pandy:

> This is most likely an elm-reactor issue. Elm-reactor wraps setTimeout and clearTimeout, but it does not wrap requestAnimationFrame. This is needed to it can catch the callbacks and re-execute them when resuming the program.
> 
> You can see that fpsWhen uses elm.setTimeout (not window.setTimeout):
> https://github.com/elm-lang/core/blob/master/src/Native/Time.js#L31
> But AnimationFrame uses window.requestAnimationFrame:
> https://github.com/jwmerrill/elm-animation-frame/blob/master/src/Native/AnimationFrame.js#L13

**Comment by jwmerrill** [comment]
I'm trying to evaluate the best way to fix this. In analogy with `fpsWhen`'s use of `setTimeout`, I think the right solution might be to
1. Add `elm.requestAnimationFrame` to the runtime's exported values as a straight pass through to `window.requestAnimationFrame`, as is done for `setTimeout` here: https://github.com/elm-lang/core/blob/8d614b595b35d1436933793cf7023393ce8274e6/src/Native/Runtime.js#L82
2. In elm-reactor, override `elm.requestAnimationFrame` with a wrapped function that allows pausing, as is done for `setTimeout` here: https://github.com/elm-lang/elm-reactor/blob/60b8dca16d5ba4a95a6a2aafe7d89a5a207639d6/frontend/debugger-implementation.js#L719
3. Switch to using `elm.requestAnimationFrame` in this library.

A couple things I'm not completely sure about:
1. The runtime itself uses `requestAnimationFrame` during rendering. I think you don't want to allow pausing that, so that usage would not be switched over to the wrapped version: https://github.com/elm-lang/core/blob/8d614b595b35d1436933793cf7023393ce8274e6/src/Native/Runtime.js#L314
2. It looks like only some of the uses (maybe only one of them?) of `setTimeout` in core use the wrapped version: https://github.com/elm-lang/core/search?utf8=%E2%9C%93&q=setTimeout Is that intentional? Also, `Time.every` uses an unwrapped `setInterval`: https://github.com/elm-lang/core/blob/79cd80e68d0fe448713b4aefdae6be80ec23675a/src/Native/Time.js#L69 Does that work properly with the reactor? How? Does it somehow involve the fact that `notify` is also wrapped by the reactor?

@laszlopandy: I'd be grateful for any input you have here.

**Comment by jwmerrill** [comment]
I understand how `every` works now--the `notify` wrapper just drops anything that comes in while the reactor is paused: https://github.com/elm-lang/elm-reactor/blob/60b8dca16d5ba4a95a6a2aafe7d89a5a207639d6/frontend/debugger-implementation.js#L785

**Comment by laszlopandy** [comment]
Yes `every` should work but there are probably some other parts of the standard library which don't work with `elm-reactor`.

Your plan to fix it looks very good. It is exactly what I would do.

**Comment by jwmerrill** [comment]
@vilterp does your work on the reactor affect the plan I outlined above? Is there somewhere I can find your work? It isn't merged in yet, is it?

**Comment by vilterp** [comment]
Hi @jwmerrill, thanks for your thought on this. My work is in the `next` branch of `elm-reactor`, plus this fork of core: https://github.com/vilterp/core/tree/new-debugger-api-support. I think your plan sounds good and should still work — we'll also need to apply this to `elm-effects`, which now uses `requestAnimationFrame` as well and currently causes bugs with pausing / time travel.

As for your second question, can't answer everything off the top of my head, but I remember moving `setInterval` to use a wrapped version.
